File: algorithms/math/gaussian-elimination/kotlin/GaussianElimination.kt
import kotlin.math.abs
import kotlin.math.roundToInt

fun gaussianElimination(arr: IntArray): Int {
    var idx = 0; val n = arr[idx++]
    val mat = Array(n) { DoubleArray(n+1) { arr[idx++].toDouble() } }
    for (col in 0 until n) {
        var maxRow = col
        for (row in col+1 until n) if (abs(mat[row][col]) > abs(mat[maxRow][col])) maxRow = row
        val tmp = mat[col]; mat[col] = mat[maxRow]; mat[maxRow] = tmp
        for (row in col+1 until n) {
            if (mat[col][col] == 0.0) continue
            val f = mat[row][col] / mat[col][col]
            for (j in col..n) mat[row][j] -= f * mat[col][j]
        }
    }
    val sol = DoubleArray(n)
    for (i in n-1 downTo 0) {
        sol[i] = mat[i][n]
        for (j in i+1 until n) sol[i] -= mat[i][j] * sol[j]
        sol[i] /= mat[i][i]
    }
    return sol.sum().roundToInt()
}

fun main() {
    println(gaussianElimination(intArrayOf(2, 1, 1, 3, 2, 1, 4)))
    println(gaussianElimination(intArrayOf(2, 1, 0, 5, 0, 1, 3)))
    println(gaussianElimination(intArrayOf(1, 2, 6)))
    println(gaussianElimination(intArrayOf(3, 1, 1, 1, 6, 0, 2, 1, 5, 0, 0, 3, 9)))
}

File: algorithms/math/gaussian-elimination/typescript/gaussianElimination.ts
export function gaussianElimination(arr: number[]): number {
    let idx = 0; const n = arr[idx++];
    const mat: number[][] = Array.from({length:n}, () => { const row: number[] = []; for (let j=0;j<=n;j++) row.push(arr[idx++]); return row; });
    for (let col=0;col<n;col++) {
        let maxRow=col;
        for (let row=col+1;row<n;row++) if (Math.abs(mat[row][col])>Math.abs(mat[maxRow][col])) maxRow=row;
        [mat[col],mat[maxRow]]=[mat[maxRow],mat[col]];
        for (let row=col+1;row<n;row++) {
            if (mat[col][col]===0) continue;
            const f=mat[row][col]/mat[col][col];
            for (let j=col;j<=n;j++) mat[row][j]-=f*mat[col][j];
        }
    }
    const sol=new Array(n).fill(0);
    for (let i=n-1;i>=0;i--) {
        sol[i]=mat[i][n];
        for (let j=i+1;j<n;j++) sol[i]-=mat[i][j]*sol[j];
        sol[i]/=mat[i][i];
    }
    return Math.round(sol.reduce((a:number,b:number)=>a+b,0));
}

console.log(gaussianElimination([2, 1, 1, 3, 2, 1, 4]));
console.log(gaussianElimination([2, 1, 0, 5, 0, 1, 3]));
console.log(gaussianElimination([1, 2, 6]));
console.log(gaussianElimination([3, 1, 1, 1, 6, 0, 2, 1, 5, 0, 0, 3, 9]));
